share/spack/setup-env.sh

```diff
@@ -251,6 +243,31 @@ function _spack_determine_shell() {
 }
 export SPACK_SHELL=$(_spack_determine_shell)
 
+# Many users will override cd to echo things to the console, which breaks
+# some commands in setup-env.sh.  We use "$_sp_builtin cd" to either
+# `command cd` (bourne shell, dash) or `builtin cd` (bash, zsh) so that
+# we don't conflict with users' custom aliases or functions unless we're
+# really doing a cd
+if [ "$SPACK_SHELL" = "bash" -o "$SPACK_SHELL" = "zsh" ]; then
+    _sp_builtin='builtin'  # bash and zsh use builtin
+else
+    _sp_builtin='command'  # sh/dash need to use command
+fi
+
+# The CDPATH variable also causes cd to echo in bash
+if [ -n "$CDPATH" ]; then
+    _sp_old_cdpath="$CDPATH"
+    unset CDPATH
+fi
+
+#
+# Find root directory and add bin to path.
+#
+_sp_share_dir=$($_sp_builtin cd "$(dirname $_sp_source_file)" && pwd)
+_sp_prefix=$($_sp_builtin cd "$(dirname $(dirname $_sp_share_dir))" && pwd)
+_spack_pathadd PATH       "${_sp_prefix%/}/bin"
+export SPACK_ROOT=${_sp_prefix}
```

Comment:
Shouldn't more of these have quotes around them? As is, I'm not sure if this will work if Spack is installed to a directory containing spaces.

Reply:
I think you would be better off if you didn't use $_sp_builtin and instead just did:

```
_sp_share_dir=$(cd "$(dirname $_sp_source_file)" > /dev/null && pwd)
```

This way you do not care what shell you are using.  This will work with bash, zsh and dash.

Reply:
That's way cleaner.  I'll refactor -- thanks!